llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp

```diff
@@ -452,7 +468,7 @@ class SIInsertWaitcnts : public MachineFunctionPass {
   // FLAT instruction.
   WaitEventType getVmemWaitEventType(const MachineInstr &Inst) const {
     assert(SIInstrInfo::isVMEM(Inst) || SIInstrInfo::isFLAT(Inst));
-    if (!ST->hasVscnt())
+    if (!ST->hasVscnt() || SIInstrInfo::isLDSDMA(Inst))
       return VMEM_ACCESS;
```

Comment:
Why does this need special casing?

Reply:
These instructions are mayStore, but the store is on the LDS side. The load side is VMEM. Without this check it incorrectly uses VS_CNT on gfx10+.

Reply:
I have also made it more clear on the source.